### PR TITLE
676 - Update OdsTextField, OdsPasswordTextField and OdsSearchTextField APIs

### PR DESCRIPTION
**Pull request by paulinea**

### CODE REVIEW

File: app/src/main/java/com/orange/ods/app/ui/components/textfields/TextField.kt [hunk 1]
```diff
@@ -148,7 +138,7 @@ fun TextFieldCodeImplementationColumn(
                     if (isSingleLine) stringRepresentation("singleLine", true)
                     if (hasTrailing) simple("trailing", "<trailing composable>")
                     if (hasCharacterCounter) {
-                        function("characterCounter", OdsComposable.OdsTextFieldCharacterCounter.name) {
+                        classInstance<OdsTextFieldCharacterCounter>("characterCounter") {
```
Comment (florentmaitre) [inline app/src/main/java/com/orange/ods/app/ui/components/textfields/TextField.kt:141]: `if (hasLeadingIcon) icon()` above should also be replaced with a `classInstance` of `OdsTextFieldIcon`.

File: lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt [hunk 2]
```diff
@@ -23,34 +23,78 @@ import androidx.compose.runtime.Composable
...
-@Composable
-@OdsComposable
-fun OdsTextFieldCharacterCounter(valueLength: Int, maxChars: Int, enabled: Boolean = true) {
-    OdsTextCaption(
-        modifier = Modifier
-            .padding(top = dimensionResource(id = R.dimen.spacing_xs), end = dimensionResource(id = R.dimen.spacing_m)),
-        text = "$valueLength/$maxChars",
-        enabled = enabled
-    )
+class OdsTextFieldCharacterCounter(private val valueLength: Int, private val maxChars: Int, private val enabled: Boolean = true) :
```
[the remark this answers is not in the document]
Comment (paulinea) [reply lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt:48]: I think I prefer `characterCount` and `maxCharacterCount` cause the class name is `OdsTextFieldCharacterCounter`

[the remark this answers is not in the document]
Comment (paulinea) [reply lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt:48]: @florentmaitre Should we add an `s` to the `character` word? `characterCount` or `charactersCount`? I never know...

[the remark this answers is not in the document]
Comment (florentmaitre) [reply lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt:48]: There is a lot of `count` in iOS APIs and they always use singular.

Also this topic confirms the singular : https://english.stackexchange.com/questions/499868/in-the-sentence-noun-count-count-of-noun-should-noun-be-plural-or-s

We can go with singular 🤓

File: lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt [hunk 3]
```diff
@@ -59,14 +103,14 @@ class OdsIconTrailing(val painter: Painter, val contentDescription: String? = nu
 internal class OdsExposedDropdownMenuTrailing(val expanded: Boolean, val enabled: Boolean) : OdsTextFieldTrailing()
```
[the remark this answers is not in the document]
Comment (paulinea) [reply lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt:103]: I don't understand what you mean. This class is internal and we only use it for dropdown menu.

File: docs/components/TextFields.md [hunk 4]
```diff
@@ -33,7 +38,7 @@ as to what information is requested for a text field.
 
 ## Variants
 
-### Text field
+### Text field [#](#text-field-)
```
[the remark this answers is not in the document]
Comment (paulinea) [reply docs/components/TextFields.md:41]: Yes, this was a try for @B3nz01d only on this documentation page ;)

File: docs/components/TextFields.md [hunk 5]
```diff
@@ -68,37 +74,73 @@ OdsTextField(
...
-```
-
-### Password text field
+##### Main OdsTextField API [#](#main-odstextfield-api-)
+
+Parameter | Default&nbsp;value | Description
+-- | -- | --
+<b>`value: String`</b> | | Input text to be shown in the text field
+<b>`onValueChange: (String) -> Unit`</b> | | Callback that is triggered when the input service updates the text. An updated text comes as a parameter of the callback.
+<b>`trailing: OdsTextFieldTrailing`</b> | | Trailing element to display at the end of the text field
```
[the remark this answers is not in the document]
Comment (paulinea) [reply docs/components/TextFields.md:101]: I think we will do that in another PR, this was a try for Benoît.

File: docs/components/TextFields.md [hunk 6]
```diff
@@ -169,4 +228,25 @@ Be careful, the limitation behavior should be managed by yourself in the `onValu
 if (text.length <= TextFieldMaxChars) {
     value = text
 }
-```
\ No newline at end of file
+```
+
+##### OdsTextFieldCharacterCounter [#](#odstextfieldcharactercounter-)
```
[the remark this answers is not in the document]
Comment (paulinea) [reply docs/components/TextFields.md:233]: Actually, this is a request from @B3nz01d to be as clear as possible in the documentation. As the `OdsTextFieldCharacterCounter` is public and used in the `OdsTextField` API, this is to add a usage documentation for the `characterCounter` parameter.

File: lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (florentmaitre) [reply lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt]: My bad, I did not see the `internal` quantifier 😞

File: lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt [hunk 7]
```diff
@@ -23,34 +23,78 @@ import androidx.compose.runtime.Composable
...
+     *
+     * @param bitmap Image bitmap of the icon.
+     * @param contentDescription The content description associated to this [OdsTextFieldIcon].
+     * @param onClick Callback invoked on icon click.
+     */
+    constructor(bitmap: ImageBitmap, contentDescription: String, onClick: (() -> Unit)? = null) : super(bitmap, contentDescription, onClick = onClick)
+
 }
 
 sealed class OdsTextFieldTrailing
```
[the remark this answers is not in the document]
Comment (florentmaitre) [reply lib/src/main/java/com/orange/ods/compose/component/textfield/OdsTextFieldsCommon.kt]: Fix is OK but should we rename `OdsTextTrailing` to `OdsTextFieldTrailingText` and `OdsIconTrailing` to `OdsTextFieldTrailingIcon` to follow the naming pattern of other components?

Also, I did not see it at first sight, but we should add several constructors for `OdsIconTrailing` (for `Bitmap` and `VectorImage`).

**Review by florentmaitre — APPROVED** [review]

**Pull request merged**